doc/tcpcarve.1
.TH "tcpcarve" 1 "November 2022" "ONICS 1.0"
.SH NAME
nftrk - Track network flows and tag packets with a flow ID.
.P
.SH SYNOPSYS
\fBtcpcarve\fP [\fB-hp\fP] [\fB-p prefix\fP] [\fIinfile\fP]
.P
.SH DESCRIPTION
\fBtcpcarve\fP is a utility to extract the application data from a
TCP stream.  \fBtcpcarve\fP expects to receive either a stream of TCP
packets from standard input or from a file all from the same TCP
connection. It reads from standard input unless the \fIinfile\fP
parameter is present in which case it reads from that file instead.
.P
The \fBtcpcarve\fP utility generates a series of numbered files.  It
creates a new file whenever it observes data in the TCP stream in a "new
direction".  That is to say, if it was extracting data sent by the
client of the connection and sees a packet from the server with data, it
closes the existing file opens a new one and starts filling it with data
until (from the server) until the client sends data again or until the
TCP connection ends.
.P
By default the files are named "data.NNNN.c2s" or "data.NNNN.s2c" where
NNNN is a monotonically increasing number regardless of the suffix of
the file.  The client-sent data goes into files with the suffix ".c2s"
and the server-sent data goes into files with the suffix ".s2c".  The
user can override the prefix of the file with the \fB-p\fP option.
.P
One
can easily merge all the from the client or server into a single file by
concatenating the files in ascending numeric order e.g.:
.nf

	cat data.*.c2s > client-data.txt

.fi
However, for request/response protocols like HTTP, the default split
should probably extract the data stream in a reasonable way.  
.P
.SH OPTIONS
.P
.IP \fB-h\fP
Print the usage of the command and exit.
.IP \fB-p prefix\fP
Use \fBprefix\fP (which may include directory assuming the directories
already exists) as the prefix for the files to generate.  If not
specified, the files will start with "data.".
.IP \fB-q\fP
Quiet mode:  do not report non-TCP packets, packets from other flows or
unexpected packets in the data stream.
.P
.SH EXAMPLES
Extract the TCP connections from a data stream and then extract the
application data from each TCP stream.
.nf

	pxtr '{tcp}' | nftrk -q | psplit
	for f in flow.* ; do
		fid=`echo $f | sed -e 's/^flow.//' -e 's/.xpkt$//'`
		tcpcarve -p "tcp-$fid." $f
	done
		

.fi
.P
.SH AVAILABILITY
This utility should be available on all supported ONICS platforms.
.P
.SH AUTHOR
Christopher A. Telfer <ctelfer@gmail.com>
.P
.SH "SEE ALSO"
nftrk(1), pml(1), psplit(1), pxtr(1)


src/tcpcarve.c
/*
 * ONICS
 * Copyright 2021
 * Christopher Adam Telfer
 *
 * tcpcarve.c -- Extract application data from a stream of TCP packets
 *
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <cat/cat.h>
#include <cat/err.h>
#include <cat/list.h>
#include <cat/optparse.h>
#include <cat/pack.h>
#include <cat/str.h>
#include "pktbuf.h"
#include "stdproto.h"
#include "tcpip_hdrs.h"


#define MAXFNAME 65536
const char *prefix = "data.";
const char *progname;
int quiet = 0;
FILE *infile;

struct clopt options[] = {
	CLOPT_I_NOARG('h', NULL, "print help"),
	CLOPT_I_STRING('p', NULL, "PREFIX",
		       "Prefix to use for each file"),
	CLOPT_I_NOARG('q', NULL,
		      "Suppress warnings for missing data or non-TCP packets"),
};
struct clopt_parser oparse =
CLOPTPARSER_INIT(options, array_length(options));


enum {
	INVALID = -1,
	C2S = 0,
	S2C = 1,
};


union addr {
	uint32_t		ip;
	struct ipv6addr		ip6;
};

struct conn_tuple {
	union addr saddr;
	union addr daddr;
	uint16_t sport;
	uint16_t dport;
	uint8_t netproto;
};

struct tcp_state {
	int seen;
	int dir;
	uint32_t nxtseq;
	int npending;
	struct list pending;
};


void usage(const char *estr)
{
	char str[4096];
	if (estr)
		fprintf(stderr, "%s\n", estr);
	optparse_print(&oparse, str, sizeof(str));
	fprintf(stderr, "usage: %s [options] [INFILE]\nOptions:\n%s\n",
		progname, str);
	exit(1);
}


void parse_args(int argc, char *argv[])
{
	int rv;
	struct clopt *opt;
	const char *fn;

	infile = stdin;
	progname = argv[0];

	optparse_reset(&oparse, argc, argv);
	while (!(rv = optparse_next(&oparse, &opt))) {
		switch (opt->ch) {
		case 'h':
			usage(NULL);
			break;
		case 'p':
			prefix = opt->val.str_val;
			break;
		case 'q':
			quiet = 1;
			break;
		}
	}
	if (rv < 0)
		usage(oparse.errbuf);

	if (rv < argc) {
		fn = argv[rv++];
		infile = fopen(fn, "r");
		if (infile == NULL)
			errsys("Error opening file %s: ", fn);
	}
}


int get_tcp(struct pktbuf *p, struct prparse **prp, struct tcph **tcp)
{
	*prp = NULL;
	*tcp = NULL;
	if ((*prp = p->layers[PKB_LAYER_XPORT]) == NULL)
		return -1;
	if ((*prp)->prid != PRID_TCP)
		return -1;
	*tcp = prp_header(*prp, p->buf, struct tcph);
	return 0;
}


void Get_tcp(struct pktbuf *p, struct prparse **prp, struct tcph **tcp)
{
	int rv = get_tcp(p, prp, tcp);
	abort_unless(rv >= 0);
}


void tcp_state_init(struct tcp_state *ts)
{
	memset(ts, 0, sizeof(*ts));
	l_init(&ts->pending);
}


#define le2pkb(le) container((le), struct pktbuf, entry)

int32_t seq_cmp(uint32_t s1, uint32_t s2)
{
	return (int32_t)(s1 - s2);
}

void tcp_state_enqueue(struct tcp_state *ts, struct pktbuf *p)
{
	struct prparse *prp;
	struct tcph *tcp;
	struct prparse *prp2;
	struct tcph *tcp2;
	struct list *node, *xnode;
	int32_t n;

	Get_tcp(p, &prp, &tcp);
	l_for_each_safe(node, xnode, &ts->pending) {
		Get_tcp(le2pkb(node), &prp2, &tcp2);
		n = seq_cmp(ntoh32(tcp->seqn), ntoh32(tcp2->seqn));
		if (n == 0) {
			pkb_free(p);
			return;
		}
		if (n > 0) {
			++ts->npending;
			l_ins(node, &p->entry);
			return;
		}
	}
	++ts->npending;
	l_enq(&ts->pending, &p->entry);
}


void tcp_state_write_pending(FILE *outfile, struct tcp_state *ts)
{
	struct pktbuf *p;
	struct prparse *prp;
	struct tcph *tcp;
	int32_t n;
	uint32_t start;
	uint32_t end;
	uint32_t off;
	uint32_t len;
	byte_t *data;
	size_t nw;

	while (!l_isempty(&ts->pending)) {
		p = le2pkb(l_head(&ts->pending));
		Get_tcp(p, &prp, &tcp);
		start = ntoh32(tcp->seqn);
		n = seq_cmp(ts->nxtseq, start);
		if (n < 0)
			return;

		l_rem(&p->entry);
		end = start + prp_plen(prp);
		if (n > 0) {
			n = seq_cmp(ts->nxtseq, end);
			if (n >= 0) {
				pkb_free(p);
				continue;
			}
			off = ts->nxtseq - start;
			data = prp_payload(prp, p->buf) + off;
			len = prp_plen(prp) - off;
		} else {
			data = prp_payload(prp, p->buf);
			len = prp_plen(prp);
		}

		nw = fwrite(data, 1, len, outfile);
		if (nw < len)
			errsys("fwrite(): ");
		pkb_free(p);
		ts->nxtseq += len;
	}
}


int extract_tuple(struct pktbuf *p, struct conn_tuple *t)
{
	struct prparse *prp;
	struct ipv4h *ip;
	struct ipv6h *ip6;
	struct tcph *tcp;

	memset(t, 0, sizeof(*t));
	if ((prp = p->layers[PKB_LAYER_NET]) == NULL)
		return -1;
	t->netproto = prp->prid;
	if (prp->prid == PRID_IPV4) {
		ip = prp_header(prp, p->buf, struct ipv4h);
		t->saddr.ip = ip->saddr;
		t->daddr.ip = ip->daddr;
	} else if (prp->prid == PRID_IPV6) {
		ip6 = prp_header(prp, p->buf, struct ipv6h);
		t->saddr.ip6 = ip6->saddr;
		t->daddr.ip6 = ip6->daddr;
	} else {
		return -1;
	}

	if ((prp = p->layers[PKB_LAYER_XPORT]) == NULL)
		return -1;
	if (prp->prid != PRID_TCP)
		return -1;
	tcp = prp_header(prp, p->buf, struct tcph);
	if (get_tcp(p, &prp, &tcp) < 0)
		return -1;
	t->sport = tcp->sport;
	t->dport = tcp->dport;

	return 0;
}


int tuple_match(struct conn_tuple *ct, struct conn_tuple *pt)
{
	struct conn_tuple rt = {0};
	if (memcmp(ct, pt, sizeof(*ct)) == 0)
		return C2S;
	rt.saddr = pt->daddr;
	rt.daddr = pt->saddr;
	rt.sport = pt->dport;
	rt.dport = pt->sport;
	rt.netproto = pt->netproto;
	if (memcmp(ct, &rt, sizeof(*ct)) == 0)
		return S2C;
	return INVALID;
}


FILE *open_outfile(ulong nfiles, int dir)
{
	char ofname[MAXFNAME];
	int rv;
	char *sfx = (dir == C2S) ? ".c2s" : ".s2c";
	FILE *of;

	rv = str_fmt(ofname, sizeof(ofname), "%s%04lu%s", prefix, nfiles, sfx);
	if (rv < 0 || rv >= MAXFNAME)
		err("filename too long!");

	of = fopen(ofname, "w");
	if (of == NULL)
		errsys("fopen(\"%s\", \"w\"): ", ofname);
	return of;
}


/* Write 0xdb through to next packet and flush contiguous packets from there */
void flush_next_pending(FILE *outfile, struct tcp_state *ts)
{
	struct pktbuf *p = le2pkb(l_head(&ts->pending));
	struct prparse *prp;
	struct tcph *tcp;
	uint32_t start;
	uint32_t len;
	size_t wlen;
	int nw = 0;
	byte_t buf[4096];

	Get_tcp(p, &prp, &tcp);
	start = ntoh32(tcp->seqn);
	abort_unless(seq_cmp(ts->nxtseq, start) < 0);
	len = start - ts->nxtseq;
	memset(buf, 0xdb, sizeof(buf));
	while (len > 0) {
		wlen = len > sizeof(buf) ? sizeof(buf) : len;
		nw = fwrite(buf, 1, wlen, outfile);
		if (nw < wlen)
			errsys("fwrite(): ");
		len -= nw;
		ts->nxtseq += nw;
	}

	tcp_state_write_pending(outfile, ts);
}


/* flush out through all enqueued packets */
void flush_outfile(FILE *outfile, struct tcp_state *ts)
{
	while (!l_isempty(&ts->pending))
		flush_next_pending(outfile, ts);
	fclose(outfile);
}


void extract_data(FILE *outfile, struct pktbuf *p, struct tcp_state *ts)
{
	tcp_state_enqueue(ts, p);
	tcp_state_write_pending(outfile, ts);
}


void carve_app_data()
{
	FILE *outfile = NULL;
	struct tcp_state c2s;
	struct tcp_state s2c;
	struct tcp_state *ts;
	int fdir = INVALID;
	int pdir;
	int rv;
	struct pktbuf *p;
	struct conn_tuple conntuple;
	struct conn_tuple pkttuple;
	struct prparse *prp;
	struct tcph *tcp;
	ulong pn = 0;
	ulong nfiles = 0;

	tcp_state_init(&c2s);
	tcp_state_init(&s2c);

	/* Find the first packet of the flow */
	while (1) {
		rv = pkb_file_read_a(&p, infile, NULL, NULL);
		if (rv < 0)
			errsys("pkb_file_read_a() of packet %lu: ", pn);
		if (rv == 0)
			return;
		++pn;
		if (pkb_parse(p) == 0 && extract_tuple(p, &conntuple) >= 0)
			break;
		if (!quiet)
			fprintf(stderr,
				"Packet %lu is not a TCP packet (skipping)\n",
				pn);
		pkb_free(p);
	}

	Get_tcp(p, &prp, &tcp);
	if (!(tcp->flags & TCPF_SYN) || (tcp->flags & TCPF_ACK)) {
		if (!quiet)
			fprintf(stderr, "Warning: first packet is not a SYN\n");
	}
	pkb_clear_parse(p);

	--pn;
	do {
		++pn;
		if (pkb_parse(p) < 0 || extract_tuple(p, &pkttuple) < 0) {
			if (!quiet)
				fprintf(stderr,
					"Packet %lu is not a TCP packet "
					"(skipping)\n", pn);
			pkb_free(p);
			continue;
		}

		pdir = tuple_match(&conntuple, &pkttuple);
		if (pdir == INVALID) {
			if (!quiet)
				fprintf(stderr,
					"Packet %lu is from a different "
					"connection skipping)\n", pn);
			pkb_free(p);
			continue;
		}

		Get_tcp(p, &prp, &tcp);

		ts = (pdir == C2S) ? &c2s : &s2c;
		if (!ts->seen) {
			ts->seen = 1;
			ts->nxtseq = ntoh32(tcp->seqn) + 1;
			if (!(tcp->flags & TCPF_SYN)) {
				if (!quiet) {
					char *origin = pdir == C2S 
						? "client" 
						: "server";
					fprintf(stderr, 
						"Warning: first packet from %s"
					        "is not a SYN\n", origin);
				}
				--ts->nxtseq;
			}
		}
		if (prp_plen(prp) <= 0) {
			pkb_free(p);
			continue;
		}
		if (!(tcp->flags & TCPF_ACK)) {
			if (!quiet)
				fprintf(stderr,
					"Data packet %lu has no ACK flag"
					"(skipping)\n", pn);
			pkb_free(p);
			continue;
		}


		if (pdir != fdir) {
			if (outfile != NULL) {
				ts = (fdir == C2S) ? &c2s : &s2c;
				flush_outfile(outfile, ts);
				outfile = NULL;
			}
			fdir = pdir;
			outfile = open_outfile(nfiles, fdir);
			++nfiles;
		}

		ts = (pdir == C2S) ? &c2s : &s2c;
		abort_unless(ts->seen);

		/* extract_data() will free the packet buffer or enqueue it */
		extract_data(outfile, p, ts);

	} while ((rv = pkb_file_read_a(&p, infile, NULL, NULL)) > 0);
	if (rv < 0)
		errsys("pkb_file_read_a() of packet %lu: ", pn);

	if (outfile != NULL) {
		ts = (fdir == C2S) ? &c2s : &s2c;
		flush_outfile(outfile, ts);
	}
}


int main(int argc, char *argv[])
{
	register_std_proto();
	pkb_init_pools(1);
	parse_args(argc, argv);
	carve_app_data();
	return 0;
}
